### PR TITLE
MAKE-813: runtime metrics collectors

### PR DESCRIPTION
**Pull request by tychoish**

### CODE REVIEW

File: metrics/metrics.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	out := &Runtime{
+		ID:        id,
+		PID:       pid,
+		Timestamp: time.Now(),
+	}
+
+	base := message.Base{}
+
+	if !opts.SkipGolang {
+		out.Golang = message.CollectGoStatsTotals().(*message.GoRuntimeInfo)
```
Comment (bsamek) [inline metrics/metrics.go]: Do you really want these to panic?

File: metrics/metrics.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+		out.Process.Base = base
+	}
+
+	return out
+
+}
+
+// NewCollectionOptions creates a valid, populated collection options
+// structure, collecting data every minute, rotating files every 24
+// hours, with 1000
```
Comment (bsamek) [inline metrics/metrics.go]: "with 300 samples"?

File: metrics/metrics.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+func (opts CollectOptions) Validate() error {
+	catcher := grip.NewBasicCatcher()
+
+	catcher.NewWhen(opts.FlushInterval < time.Millisecond,
+		"flush interval must be greater than a millisecond")
+	catcher.NewWhen(opts.CollectionInterval < time.Millisecond,
+		"collection interval must be greater than a millisecond")
+	catcher.NewWhen(opts.CollectionInterval > opts.FlushInterval,
+		"collection interval must be smaller than flush interval")
+	catcher.NewWhen(opts.SampleCount < 10, "sample count must be greater than 10")
```
Comment (bsamek) [inline metrics/metrics.go]: nit `<=`

File: metrics/metrics.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+		outputCount++
+
+		file, err = os.Create(fmt.Sprintf("%s.%d", opts.OutputFilePrefix, outputCount))
+		if err != nil {
+			return errors.Wrap(err, "problem creating subsequent file")
+		}
+
+		collector = ftdc.NewStreamingCollector(opts.SampleCount, file)
+		flushTimer.Reset(opts.FlushInterval)
```
Comment (bsamek) [inline metrics/metrics.go]: style nit: I'd argue for moving this below `case <-flushTimer.C`for clarity.

File: metrics/metrics.go (hunk 1 above)
Comment (tychoish) [reply metrics/metrics.go]: I'm ok with it.

Comment (tychoish) [reply metrics/metrics.go]: (the tests would catch it.)

**Pull request merged**